File: src/main/java/ekkel/book/generics/DogsAndRobots.cpp
#include <iostream>
using namespace std;

class Dog {
public:
	void speak() {
		cout << "Gav!" << endl;
	}
	void sit() {
		cout << "Sitting" << endl;
	}
	void reproduce() {}
};

class Robot {
public:
	void speak() {
		cout << "Click!" << endl;
	}
	void sit() {
		cout << "Clank!" << endl;
	}
	void oilChange() {}
};

class Mime {
public:
	void sit() {
		cout << "Pretend to sit" << endl;
	}
	void pushInvisibleWall() {}
};

template<class T> void perform(T anything) {
	anything.speak();
	anything.sit();
}

int main() {
	Dog d;
	Robot r;
	Mime m;
	
	perform(d);
	perform(r);
	//perform(m);
}

File: src/main/java/ekkel/book/generics/DogsAndRobots.py
class Dog:
    def speak(self):
        print('Arf!')

    def sit(self):
        print("Sitting")

    def reproduce(self):
        pass


class Robot:
    def speak(self):
        print("Click!")

    def sit(self):
        print("Clank!")

    def oil_change(self):
        pass


class Mime:
    def sit(self):
        print("Pretend to sit")

    def push_invisible_wall(self):
        pass


def perform(anything):
    anything.speak()
    anything.sit()


a = Dog()
b = Robot()
c = Mime()

perform(a)
perform(b)
# perform(c)
